### PR TITLE
CRS-2186-se20-offence-validation

### PR DESCRIPTION
**Pull request by mg-moj**
Add before calc validation to prevent release dates being calculated for Sentencing Act 2020 (SA20) offences that have been entered in NOMIS with an offence date prior to 1/12/20.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/validation/PreCalculationValidationService.kt [hunk 1]
```diff
@@ -12,6 +12,7 @@ import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.external.Sent
 import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.external.SentenceCalculationType
 import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.model.external.SentenceCalculationType.Companion.from
 import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.service.ImportantDates.PCSC_COMMENCEMENT_DATE
+import uk.gov.justice.digital.hmpps.calculatereleasedatesapi.service.ImportantDates.SHPO_OFFENCE_COMMENCEMENT_DATE
```
Comment (srowlands-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/validation/PreCalculationValidationService.kt:15]: Could you change this to be SENTENCING_ACT_2020_COMMENCEMENT or something, as it's not SHPO related anymore, SEC_91_END_DATE could then also be updated to use the same more generic date

Comment (mg-moj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/validation/PreCalculationValidationService.kt:15]: Yes that makes more sense in light of the changes.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/validation/PreCalculationValidationService.kt [hunk 2]
```diff
@@ -91,13 +98,37 @@ class PreCalculationValidationService(
...
     return messages
   }
 
+  fun validateSe20Offences(data: PrisonApiSourceData): List<ValidationMessage> {
+    val invalidOffences = data.sentenceAndOffences.filter {
+      it.offence.offenceCode.startsWith("SE20") &&
+        it.offence.offenceStartDate?.isBefore(SHPO_OFFENCE_COMMENCEMENT_DATE) ?: false
+    }
+
+    return if (invalidOffences.count() == 1) {
```
Comment (srowlands-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/calculatereleasedatesapi/validation/PreCalculationValidationService.kt:115]: `.size`?

**Comment by srowlands-moj** [comment]
Are you going to add some tests back once Scott has updated the ticket?

**Review by srowlands-moj — APPROVED** [review]
LGTM

**Pull request merged**